src/main/java/carpet/script/utils/ShapesRenderer.java

```diff
@@ -233,6 +288,179 @@ public void promoteWith(RenderedShape<?> rshape)
+                    //client.getBlockRenderer().renderSingleBlock(blockState, matrices, immediate, light,
+                    //        OverlayTexture.NO_OVERLAY);
+                    var bakedModel = client.getBlockRenderer().getBlockModel(blockState);
+                    int color = client.getBlockColors().getColor(blockState,client.level,blockPos,0);
+                    //dont know why there is a 0. 
+                    //see https://github.com/senseiwells/EssentialClient/blob/4db1f291936f502304791ee323f369c206b3021d/src/main/java/me/senseiwells/essentialclient/utils/render/RenderHelper.java#L464
+                    float red = (color >> 16 & 0xFF) / 255.0F;
+			    	float green = (color >> 8 & 0xFF) / 255.0F;
+			    	float blue = (color & 0xFF) / 255.0F;
+                    client.getBlockRenderer().getModelRenderer().renderModel(matrices.last(), immediate.getBuffer(ItemBlockRenderTypes.getRenderType(blockState, false)), blockState, bakedModel, red, green, blue, light, OverlayTexture.NO_OVERLAY);
+                }
+
+                // draw the block`s entity part
+                if (BlockEntity == null) {
+                    if (blockState.getBlock() instanceof EntityBlock eb) {
+                        BlockEntity = eb.newBlockEntity(blockPos, blockState);
+                        if (BlockEntity != null) {
+                            BlockEntity.setLevel(client.level);
+                            if (shape.blockEntity != null)
+                                BlockEntity.load(shape.blockEntity);
+                        }
+                    }
+                }
+                if (BlockEntity instanceof ShulkerBoxBlockEntity sbBlockEntity){
+                    sbrender(sbBlockEntity, partialTick,
+                    matrices, immediate, light, OverlayTexture.NO_OVERLAY);
+                }else
+                if (BlockEntity != null && client.getBlockEntityRenderDispatcher().getRenderer(BlockEntity) != null) {
+                    client.getBlockEntityRenderDispatcher().getRenderer(BlockEntity).render(BlockEntity, partialTick,
+                            matrices, immediate, light, OverlayTexture.NO_OVERLAY);
+                }
+            } else {
+                // draw item
+                if (shape.item != null)
+                    client.getItemRenderer().renderStatic(shape.item, shape.item_transform_type, light,
+                            OverlayTexture.NO_OVERLAY, matrices, immediate, (int) shape.key());
+            }
+            matrices.popPose();
+            immediate.endBatch();
+            RenderSystem.disableCull();
+            RenderSystem.disableDepthTest();
+            RenderSystem.depthMask(false);
+
+        }
+
+        @Override
+        public boolean stageDeux() {
+            return true;
+        }
+
+        public void sbrender(ShulkerBoxBlockEntity shulkerBoxBlockEntity, float f, PoseStack poseStack, MultiBufferSource multiBufferSource, int i, int j) {
```

Comment:
if you copy - paste- modify this method from vanilla code, make sure to include the reference in a comment to allow for simpler tracking in case that stops working or becomes weird

Reply:
i added a comment to mark its origin